File: src/main/java/com/github/simiacryptus/aicoder/actions/MarkdownListAction.java
package com.github.simiacryptus.aicoder.actions;

import com.github.simiacryptus.aicoder.config.AppSettingsState;
import com.github.simiacryptus.aicoder.openai.CompletionRequest;
import com.github.simiacryptus.aicoder.util.psi.PsiUtil;
import com.github.simiacryptus.aicoder.util.StringTools;
import com.github.simiacryptus.aicoder.util.UITools;
import com.intellij.openapi.actionSystem.AnAction;
import com.intellij.openapi.actionSystem.AnActionEvent;
import com.intellij.openapi.actionSystem.CommonDataKeys;
import com.intellij.openapi.editor.Caret;
import com.intellij.openapi.editor.Document;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiFile;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Arrays;
import java.util.List;
import java.util.Objects;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import static com.github.simiacryptus.aicoder.util.UITools.insertString;

public class MarkdownListAction extends AnAction {

    @Override
    public void update(@NotNull AnActionEvent e) {
        e.getPresentation().setEnabledAndVisible(isEnabled(e));
        super.update(e);
    }

    private static boolean isEnabled(@NotNull AnActionEvent e) {
        return null != getMarkdownListParams(e);
    }

    @Nullable
    public static MarkdownListParams getMarkdownListParams(@NotNull AnActionEvent e) {
        Caret caret = e.getData(CommonDataKeys.CARET);
        if (null == caret) return null;
        PsiFile psiFile = e.getData(CommonDataKeys.PSI_FILE);
        if (null == psiFile) return null;
        PsiElement list = PsiUtil.getSmallestIntersecting(psiFile, caret.getSelectionStart(), caret.getSelectionEnd(), "MarkdownListImpl");
        if (null == list) return null;
        return new MarkdownListParams(caret, list);
    }

    @Override
    public void actionPerformed(@NotNull AnActionEvent event) {
        MarkdownListParams markdownListParams = getMarkdownListParams(event);
        AppSettingsState settings = AppSettingsState.getInstance();
        List<CharSequence> items = StringTools.trim(PsiUtil.getAll(Objects.requireNonNull(markdownListParams).list, "MarkdownListItemImpl")
                .stream().map(item -> PsiUtil.getAll(item, "MarkdownParagraphImpl").get(0).getText()).collect(Collectors.toList()), 10, false);
        CharSequence indent = UITools.getIndent(markdownListParams.caret);
        CharSequence n = Integer.toString(items.size() * 2);
        int endOffset = markdownListParams.list.getTextRange().getEndOffset();
        String listPrefix = "* ";
        CompletionRequest completionRequest = settings.createTranslationRequest()
                .setInstruction(UITools.getInstruction("List " + n + " items"))
                .setInputType("instruction")
                .setInputText("List " + n + " items")
                .setOutputType("list")
                .setOutputAttrute("style", settings.style)
                .buildCompletionRequest()
                .appendPrompt(items.stream().map(x2 -> listPrefix + x2).collect(Collectors.joining("\n")) + "\n" + listPrefix);
        Document document = event.getRequiredData(CommonDataKeys.EDITOR).getDocument();
        UITools.redoableRequest(completionRequest, "", event, newText -> transformCompletion(markdownListParams, indent, listPrefix, newText), newText -> insertString(document, endOffset, newText));
    }

    @NotNull
    private static String transformCompletion(MarkdownListParams markdownListParams, CharSequence indent, String listPrefix, CharSequence complete) {
        List<CharSequence> newItems = Arrays.stream(complete.toString().split("\n")).map(String::trim)
                .filter(x1 -> x1.length() > 0).map(x1 -> StringTools.stripPrefix(x1, listPrefix)).collect(Collectors.toList());
        String strippedList = Arrays.stream(markdownListParams.list.getText().split("\n"))
                .map(String::trim).filter(x -> x.length() > 0).collect(Collectors.joining("\n"));
        String bulletString = Stream.of("- [ ] ", "- ", "* ")
                .filter(strippedList::startsWith).findFirst().orElse("1. ");
        CharSequence itemText = indent + newItems.stream().map(x -> bulletString + x)
                .collect(Collectors.joining("\n" + indent));
        return "\n" + itemText;
    }

    public static class MarkdownListParams {
        public final Caret caret;
        public final PsiElement list;

        private MarkdownListParams(Caret caret, PsiElement list) {
            this.caret = caret;
            this.list = list;
        }

    }
}

File: src/main/kotlin/aicoder/actions/markdown/MarkdownListAction.kt
package aicoder.actions.markdown

import aicoder.actions.BaseAction
import com.intellij.openapi.actionSystem.ActionUpdateThread
import com.intellij.openapi.actionSystem.AnActionEvent
import com.intellij.openapi.actionSystem.CommonDataKeys
import com.intellij.openapi.application.ApplicationManager
import com.intellij.openapi.diagnostic.Logger
import com.intellij.openapi.progress.ProgressIndicator
import com.intellij.openapi.project.Project
import com.simiacryptus.aicoder.config.AppSettingsState
import com.simiacryptus.aicoder.util.UITools
import com.simiacryptus.aicoder.util.UITools.getIndent
import com.simiacryptus.aicoder.util.UITools.insertString
import com.simiacryptus.aicoder.util.psi.PsiUtil.getAll
import com.simiacryptus.aicoder.util.psi.PsiUtil.getSmallestIntersecting
import com.simiacryptus.jopenai.models.chatModel
import com.simiacryptus.jopenai.proxy.ChatProxy
import com.simiacryptus.util.StringUtil
import java.awt.Component
import javax.swing.JOptionPane

/**
 * Action that extends markdown lists by generating additional items using AI.
 * Supports bullet lists and checkbox lists.
 */
class MarkdownListAction : BaseAction() {
  private val log = Logger.getInstance(MarkdownListAction::class.java)
  private lateinit var progress: ProgressIndicator

  data class ListConfig(
    val itemCount: Int = 0,
    val temperature: Double = AppSettingsState.instance.temperature
  )

  /**
   * Gets configuration for list generation
   */
  private fun getConfig(project: Project?): ListConfig? {
    return try {
      ListConfig(
        itemCount = UITools.showInputDialog(
          project as? Component,
          "How many new items to generate?",
          "Generate List Items",
          JOptionPane.QUESTION_MESSAGE
        )?.let { Integer.parseInt(it.toString()) } ?: return null
      )
    } catch (e: Exception) {
      log.warn("Failed to get configuration", e)
      null
    }
  }

  override fun getActionUpdateThread(): ActionUpdateThread = ActionUpdateThread.BGT

  interface ListAPI {
    fun newListItems(
      items: List<String?>?,
      count: Int,
    ): Items

    data class Items(
      val items: List<String?>? = null,
    )
  }

  val proxy: ListAPI
    get() {
      val chatProxy = ChatProxy(
        clazz = ListAPI::class.java,
        api = api,
        model = AppSettingsState.instance.smartModel.chatModel(),
        temperature = AppSettingsState.instance.temperature,
        deserializerRetries = 5
      )
      chatProxy.addExample(
        returnValue = ListAPI.Items(
          items = listOf("Item 4", "Item 5", "Item 6")
        )
      ) {
        it.newListItems(
          items = listOf("Item 1", "Item 2", "Item 3"),
          count = 6
        )
      }
      return chatProxy.create()
    }

  override fun handle(e: AnActionEvent) {
    try {
      val caret = e.getData(CommonDataKeys.CARET) ?: return
      val psiFile = e.getData(CommonDataKeys.PSI_FILE) ?: return
      val project = e.project ?: return
      val config = getConfig(project) ?: return
      val list =
        getSmallestIntersecting(psiFile, caret.selectionStart, caret.selectionEnd, "MarkdownListImpl") ?: return
      val items = StringUtil.trim(
        getAll(list, "MarkdownListItemImpl")
          .map {
            val all = getAll(it, "MarkdownParagraphImpl")
            if (all.isEmpty()) it.text else all[0].text
          }.toList(), 10, false
      )
      progress.fraction = 0.4
      progress.text = "Generating new items..."
      val indent = getIndent(caret)
      val endOffset = list.textRange.endOffset
      val bulletTypes = listOf("- [ ] ", "- ", "* ")
      val document = (e.getData(CommonDataKeys.EDITOR) ?: return).document
      val rawItems = items.map(CharSequence::trim).map {
        val bulletType = bulletTypes.find(it::startsWith)
        if (null != bulletType) StringUtil.stripPrefix(it, bulletType).toString()
        else it.toString()
      }

      UITools.redoableTask(e) {
        var newItems: List<String?>? = null
        progress.isIndeterminate = false
        progress.fraction = 0.2
        progress.text = "Analyzing existing items..."
        UITools.run(
          e.project, "Generating New Items", true
        ) {
          newItems = proxy.newListItems(
            rawItems,
            config.itemCount
          ).items
          progress.fraction = 0.8
          progress.text = "Formatting results..."
        }
        var newList = ""
        ApplicationManager.getApplication().runReadAction {
          val strippedList = list.text.split("\n")
            .map(String::trim).filter(String::isNotEmpty)
            .joinToString("\n")
          val bulletString = bulletTypes.find(strippedList::startsWith) ?: "1. "
          newList = newItems?.joinToString("\n") { indent.toString() + bulletString + it } ?: ""
        }
        UITools.writeableFn(e) {
          insertString(document, endOffset, "\n" + newList)
        }
      }

    } catch (ex: Exception) {
      log.error("Failed to generate list items", ex)
      UITools.showErrorDialog(
        "Failed to generate list items: ${ex.message}",
        "Error"
      )
    }
  }

  override fun isEnabled(e: AnActionEvent): Boolean {
    val caret = e.getData(CommonDataKeys.CARET) ?: return false
    val psiFile = e.getData(CommonDataKeys.PSI_FILE) ?: return false
    getSmallestIntersecting(psiFile, caret.selectionStart, caret.selectionEnd, "MarkdownListImpl") ?: return false
    return super.isEnabled(e)
  }

}
